applications/luci-app-shadowsocks-libev/htdocs/luci-static/resources/view/shadowsocks-libev/rules.js

```diff
@@ -0,0 +1,117 @@
+'use strict';
+'require uci';
+'require form';
+'require shadowsocks-libev as ss';
+
+var conf = 'shadowsocks-libev';
+
+function src_dst_option(s /*, ... */) {
+	var o = s.taboption.apply(s, L.varargs(arguments, 1));
+	o.datatype = 'or(ipaddr,cidr)';
+}
+
+return L.view.extend({
+	load: function() {
+		return uci.load(conf)
+			.then(function() {
+				if (!uci.get_first(conf, 'ss_rules')) {
+					uci.set(conf, uci.add(conf, 'ss_rules', 'ss_rules'), 'disabled', '1');
+				}
+			});
+	},
+	render: function() {
+		var m, s, o;
+
+		m = new form.Map(conf, _('Redir Rules'),
+			_('On this page you can configure how traffics are to be \
+				forwarded to ss-redir instances. \
+				If enabled, packets will first have their src ip addresses checked \
+				against <em>Src ip/net bypass</em>, <em>Src ip/net forward</em>, \
+				<em>Src ip/net checkdst</em> and if none matches <em>Src default</em> \
+				will give the default action to be taken. \
+				If the prior check results in action <em>checkdst</em>, packets will continue \
+				to have their dst addresses checked.'));
+
+		s = m.section(form.NamedSection, 'ss_rules', 'ss_rules');
+		s.tab('general', _('General Settings'));
+		s.tab('src', _('Source Settings'));
+		s.tab('dst', _('Destination Settings'));
+
+		s.taboption('general', form.Flag, 'disabled', _('Disable'));
+		//ss.option_install_package(s, 'general'); // FIXME
+
+		o = s.taboption('general', form.ListValue, 'redir_tcp',
+			_('ss-redir for TCP'));
+		ss.values_redir(o, 'tcp');
+		o = s.taboption('general', form.ListValue, 'redir_udp',
+			_('ss-redir for UDP'));
+		ss.values_redir(o, 'udp');
+
+		o = s.taboption('general', form.ListValue, 'local_default',
+			_('Local-out default'),
+			_('Default action for locally generated TCP packets'));
+		ss.values_actions(o);
+		o = s.taboption('general', form.DynamicList, 'ifnames',
+			_('Ingress interfaces'),
+			_('Only apply rules on packets from these network interfaces'));
+		//ss.values_ifnames(o); // FIXME DeviceSelect with DynamicList?
```

Comment:
Should we implement a `DynamicDeviceSelect` in widget.js?

Reply:
Ping @jow-

Reply:
Is this expecting logical interface (`lan`, `wan`) or physical ones (`eth0`, `br-lan`) ?

Maybe consider using `widgets.NetworkSelect` (or `widgets.DeviceSelect`) instead, with `.multiple = true`.

Reply:
It is physical.
![image](https://user-images.githubusercontent.com/12028138/67673049-1d267580-f9b4-11e9-9026-3df889ddc19c.png)